tests/scalar_advection/scalar_advection_adaptive.py

```diff
@@ -0,0 +1,133 @@
+cone = 1.0 - min_value(sqrt(pow(x-cone_x0, 2) + pow(y-cone_y0, 2))/cyl_r0, 1.0)
+slot_cyl = conditional(sqrt(pow(x-cyl_x0, 2) + pow(y-cyl_y0, 2)) < cyl_r0,
+                       conditional(And(And(x > slot_left, x < slot_right), y < slot_top),
+                       0.0, 1.0), 0.0)
+
+# We then declare the inital condition of :math:`q` to be the sum of these fields.
+# Furthermore, we add 1 to this, so that the initial field lies between 1 and 2,
+# rather than between 0 and 1.  This ensures that we can't get away with
+# neglecting the inflow boundary condition.  We also save the initial state so
+# that we can check the :math:`L^2`-norm error at the end. ::
+
+q_init = Function(Q).interpolate(1.0 + bell + cone + slot_cyl)
+q = Function(Q).assign(q_init)
+
+# We declare the output filename, and write out the initial condition. ::
+
+outfile = VTKFile("CG_SUadv_adaptive_q.pvd")
+outfile.write(q)
+
+u_outfile = VTKFile("CG_SUadv_adaptive_u.pvd")
+u_outfile.write(u)
+# We will run for time :math:`2\pi`, a full rotation.  The initial timestep is set
+# as a starting point for the adaptive stepper, which will adjust dt based on error
+# estimates.  Finally, we define the inflow boundary condition, :math:`q_\mathrm{in}`.
+# In general, this would be a ``Function``, but here we just use a ``Constant`` value. ::
+
+T = 2*pi
+dt = T/600.0  # Initial timestep - adaptive stepper will adjust this
+q_in = Constant(1.0)
+
+# Use G-ADOPT's GenericTransportSolver to advect the tracer with adaptive timestepping.
+# We use IrksomeRadauIIA which has an embedded error estimator for adaptive control.
+# We only include an advection term and apply weak boundary conditions on inflow regions.
+bc_in = {"q": q_in}
+bcs = {1: bc_in, 2: bc_in, 3: bc_in, 4: bc_in}
+eq_attrs = {"u": u}
+adv_solver = GenericTransportSolver(
+    "advection", q, dt, IrksomeRadauIIA,
+    eq_attrs=eq_attrs, bcs=bcs, su_advection=True,
+    timestepper_kwargs={
+        'order': 3,  # RadauIIA order
+        'adaptive_parameters': {
+            'tol': 1e-3,        # Error tolerance per step
+            'dtmin': 1e-6,      # Minimum allowed dt
+            'dtmax': T/100.0,   # Maximum allowed dt (reasonable fraction of total time)
+            'KI': 1/15,         # Integration gain
+            'KP': 0.13          # Proportional gain
+        }
+    }
+)
+
+# Get nubar (additional SU diffusion) for plotting
+nubar = Function(Q).interpolate(adv_solver.equation.su_nubar)
+nubar_outfile = VTKFile("CG_SUadv_adaptive_nubar.pvd")
+nubar_outfile.write(nubar)
+
+# Here is the time stepping loop with adaptive timestepping, with an output every 20 steps.
+# The timestep dt will be automatically adjusted by the adaptive stepper based on error estimates.
+t = 0.0
+step = 0
+dt_values = []  # Store all timestep values for testing
+while t < T:
+    # Set maximum dt to prevent overshooting final time
+    adv_solver.ts.stepper.dt_max = T - t
```

Comment:
```suggestion
    if T - t < float(adv_solver.ts.dt_reference):
        # Set maximum dt to prevent overshooting final time
        adv_solver.ts.stepper.dt_max = T - t
```